react-components/src/components/Reveal3DResources/ResourceUpdater.ts

```diff
@@ -0,0 +1,330 @@
+      newModelWithAddOptionsPromises,
+      keptModelsWithOptions
+    );
+  }
+
+  public async updateCommonStyling(
+    instanceStyling: InstanceStylingGroup[],
+    defaultStyling: DefaultResourceStyling
+  ): Promise<void> {
+    this._defaultResourceStyling = defaultStyling;
+    this._instanceStyling = instanceStyling;
+
+    this._image360StylingHandler.setCommonStyling(defaultStyling.image360, instanceStyling);
+
+    const models = await this._pendingModelsPromise;
+    models.forEach(async (model) => {
+      await this.applyStylingAndTransform(model);
+    });
+  }
+
+  private clearAllModels(): void {
+    this._viewer.models.forEach((model) => {
+      this._viewer.removeModel(model);
+    });
+    this._viewer.get360ImageCollections().forEach((collection) => {
+      this._viewer.remove360ImageSet(collection);
+    });
+  }
+
+  private updateKeptModels(
+    oldModels: AddOptionsWithModel[],
+    newOptions: AddResourceOptions[]
+  ): void {
+    oldModels.forEach((oldModel) => {
+      const correspondingNewOption = newOptions.find((newOption) =>
+        isSameModel(oldModel.addOptions, newOption)
+      );
+
+      assert(correspondingNewOption !== undefined);
+    });
+  }
+
+  private async constructNewModelPromises(
+    newModelPromises: Array<Promise<AddOptionsWithModel>>,
+    keptModels: AddOptionsWithModel[]
+  ): Promise<AddOptionsWithModel[]> {
+    const allModels = (await Promise.all(newModelPromises)).concat(keptModels);
+
+    await this._image360StylingHandler.setCollections(allModels);
+    allModels.forEach(async (model) => {
+      await this.applyStylingAndTransform(model);
+    });
+    return allModels;
+  }
+
+  private removeResources(
+    resources: AddResourceOptions[],
+    currentModels: AddOptionsWithModel[]
+  ): AddOptionsWithModel[] {
+    const currentModelsSet = new Set(currentModels);
+    resources.forEach((resourceAddOptions) => {
+      const resourceAndModel = [...currentModelsSet].find((model) =>
+        isSameModel(model.addOptions, resourceAddOptions)
+      );
+
+      assert(resourceAndModel !== undefined);
+
+      currentModelsSet.delete(resourceAndModel);
+
+      if (is3dModelOptions(resourceAddOptions)) {
+        this._viewer.removeModel(resourceAndModel.model as CogniteModel);
+      } else if (is360ImageCollectionOptions(resourceAddOptions)) {
+        this._viewer.remove360ImageSet(resourceAndModel.model as Image360Collection);
+      } else {
+        throw Error('Tried to remove Model of unknown type');
+      }
+    });
+
+    return [...currentModelsSet];
+  }
+
+  private addResources(resources: AddResourceOptions[]): Array<Promise<AddOptionsWithModel>> {
+    return resources.map(async (addOptions) => {
+      try {
+        const addedModel = await (async () => {
+          if (is3dModelOptions(addOptions)) {
+            const addedModel = await this._viewer.addModel(addOptions);
+            if (addedModel instanceof CogniteCadModel) {
```

Comment:
It would be nice to have a function `isCadModel()` and `ìsPointCloudodel()` and reuse the across `react-components`

Reply:
I could just create a function that uses `instanceof` inside, but I think it's an unnecessary abstraction. I can do it if you feel strongly about it